### PR TITLE
return blob handles while offline

### LINKED ISSUE #8246: offline blob support
related: #7594
While stashed ops have been implemented, there is no blob support. Currently, when `ContainerRuntime.uploadBlob()` is called, the promise returned will not resolve until the blob has been uploaded and a `BlobAttach` op has round-tripped. This ensures that it is not possible to attach a blob handle to the document which refers to a blob that is not present in storage because it was not successfully uploaded, or was GCed before the `BlobAttach` op was sent. However, this means that `uploadBlob()` will never resolve when we are offline.

We need to support "uploading" blobs while offline, getting blob handles, making further changes, stashing these changes, reapplying them in a new container, possibly making further changes while still offline, and finally, upon reconnection, uploading the blobs and resubmitting the stashed ops. In this scenario, we cannot rely on waiting for the `BlobAttach` op to roundtrip, but we must still ensure that we never attach an invalid handle, as it will result in a broken document.

A (partial) solution is implemented in #8159. In this PR, when a blob is uploaded while offline, `uploadBlob()` returns a handle that uses a local ID, which `BlobManager` can redirect to the storage ID once the blob is uploaded (leveraging work already done for detached container blob support). Once we reconnect, when we see a `BlobAttach` op in the resubmit queue in `PendingStateManager`, we block resubmission of any further ops until the blob has been uploaded. It should be safe in this case not to wait for the `BlobAttach` op to round-trip, since we know if it doesn't go through, nothing past it will go through either, and we can make sure upon any subsequent re-resubmission that the blob has not yet been GCed (see #6469). 

This solution is not complete as implemented in the PR since blocking resubmission requires making `PendingStateManager.replayPendingStates()` async. This breaks the assumption in `replayPendingStates()` that all pending ops will be resubmitted atomically without any new ops being submitted. The fundamental problem is that we need to block submission of new ops while allowing submission of resubmitted ops, but it's not easy to distinguish between the two since they both go through `ContainerRuntime.submit()`. If we can distinguish resubmitted ops from newly subitted ops, we can keep them separate in `PendingStateManager`, which should solve the problem.

Another solution, similar to how blobs in detached containers work, would be to upload all of the blobs before emitting "connected" event. This would ensure that all ops can be resubmitted safely.

One downside to the approach above is that it relies on the redirect table in `BlobManager`, which must be persisted in future summaries. There is no mechanism to ever remove entries from this table, so it will grow monotonically. A new entry must be added for every blob passed to `uploadBlob()` while offline, or if we want to treat online and offline blob uploads identically, a new entry must be added for all blob uploads.

### PR DESCRIPTION
**Pull request by wes-carlson**
#7594
Close #8246
Fixes [AB#54](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/54)
Return a blob handle with local ID (guid) when not connected, or upon disconnect while waiting for blob upload. These blobs are tracked by BlobManager, and runtime will delay transitioning to "connected" state until they have been uploaded. This ensures ops referencing these handles are not submitted until the blobs have been uploaded and a BlobAttach op referencing them has been submitted.

### CODE REVIEW

File: packages/runtime/container-runtime/src/blobManager.ts [hunk 1]
```diff
@@ -267,24 +410,22 @@ export class BlobManager {
                 this.redirectTable.delete(blobId);
                 continue;
             }
-            this.blobIds.delete(blobId);
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/blobManager.ts:270]: Why delete this? We still want to delete the blobs that are unused right?

File: packages/test/test-end-to-end-tests/src/test/gc/gcStats.spec.ts [hunk 2]
```diff
@@ -121,6 +121,7 @@ describeFullCompat("Garbage Collection Stats", (getTestObjectProvider) => {
         // Upload 2 attachment blobs and store their handles to mark them referenced.
         const blob1Contents = "Blob contents 1";
         const blob2Contents = "Blob contents 2";
+        await provider.ensureSynchronized();
```
Comment (agarwal-navin) [inline packages/test/test-end-to-end-tests/src/test/gc/gcStats.spec.ts:124]: Why is this required? What changed?

File: packages/runtime/container-runtime/src/blobManager.ts (hunk 1 above)
Comment (wes-carlson) [reply packages/runtime/container-runtime/src/blobManager.ts:270]: I removed `blobIds` and instead I'm adding storage ID -> storage ID entries to the redirect table and using the values (`.values()`) of the redirect table as the list of attached storage IDs. I haven't looked closely at the GC stuff but I don't think it should need to change too much.

File: packages/test/test-end-to-end-tests/src/test/gc/gcStats.spec.ts (hunk 2 above)
Comment (wes-carlson) [reply packages/test/test-end-to-end-tests/src/test/gc/gcStats.spec.ts:124]: Runtime will now return blob handles while disconnected and add a new local ID to the redirect table, changing the stats. So we have to make sure we are connected before uploading the blobs.

File: packages/runtime/container-runtime/src/blobManager.ts (hunk 1 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/blobManager.ts:270]: So, you are saying that redirectTable has all the ids. The keys are all localIds + all storageIds and values are all storageIds.    In that case, the redirectTable should be used as the source of truth for writing the ids in summary right. However, I see that in `summarize` we write `this.storageIds` if its not empty. What this means is that just deleting ids from redirect table won't be enough here. We will have to delete it from `this.storageIds` as well.

Consider this scenario:
- A blob is uploaded in attached state. It is added to `this.storageIds` as `blob1`. It it also added to redirect table as `blob1` -> `blob1`.
- When this blob is unreferenced and deleted, we remove it from the redirect table.
- When writing blob manager's summary, `blob1` still exists in `this.storageIds` and will be written into summary.

File: packages/runtime/container-runtime/src/blobManager.ts [hunk 3]
```diff
@@ -267,24 +410,22 @@ export class BlobManager {
...
         }
     }
 
     public summarize(): ISummaryTreeWithStats {
-        // If we have a redirect table it means the container is about to transition to "Attaching" state, so we need
-        // to return an actual snapshot containing all the real storage IDs we know about.
-        const attachingOrAttached = !!this.redirectTable || this.runtime.attachState !== AttachState.Detached;
-        const blobIds = attachingOrAttached ? this.blobIds : this.detachedBlobIds;
+        // if storageIds is empty, it means we are detached and have only local IDs, or that there are no blobs attached
+        const blobIds = this.storageIds.size > 0 ? Array.from(this.storageIds) : Array.from(this.redirectTable.keys());
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/blobManager.ts:418]: It looks like we add all storageIds to this.redirectTable as storageId:storageId pairs. So, we can simply always just use `this.redirectTable.keys()` which will include both local and storageIds.

Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/blobManager.ts:418]: In the current implementation, any lookup for a storageId has to be done in both redirect table and storageIds. For example, when GC is deleting an id, it has to do in both these places. If we have a single source of truth and use that exclusively, the look up only has to be done in one place.

File: packages/test/test-end-to-end-tests/src/test/gc/gcStats.spec.ts (hunk 2 above)
Comment (agarwal-navin) [reply packages/test/test-end-to-end-tests/src/test/gc/gcStats.spec.ts:124]: ensureSynchronized doesn't wait for connected state. It may or may not happen because of the different waits. You should explicitly wait for connected state. See `ensureContainerConnected` that is added to some tests.

File: packages/runtime/container-runtime/src/blobManager.ts (hunk 1 above)
Comment (wes-carlson) [reply packages/runtime/container-runtime/src/blobManager.ts:270]: `this.storageIds` is a getter, it's sourced from the redirect table

**Comment by wes-carlson** [comment]
superseded by #10621
